Distribution/Server/Features/ReverseDependencies/State.hs

```diff
@@ -13,406 +15,203 @@ import Distribution.PackageDescription
+    let index = PackageIndex.insert pkgid pindex
+        rd' | revSize revs < Bimap.size nodemap = arrDouble revs | otherwise = revs
+        npm = Bimap.tryInsert pkgid (Bimap.size nodemap) nodemap
+        rd = insEdges (map (\d -> (npm ! d, [npm ! pkgid])) deps) rd'
+    in ri {
         duplicatedIndex = index,
         reverseDependencies = rd,
-        flattenedReverse = packageNameClosure rd
-    }, rec)
-
-removePackage :: PackageId -> CombinedDeps
-              -> ReverseIndex -> (ReverseIndex, Map PackageName [Version])
-removePackage pkgid deps revs =
-    let index = PackageIndex.deletePackageId pkgid (duplicatedIndex revs)
-        (rd, rec) = unregisterPackage (getAllVersions index) pkgid deps (reverseDependencies revs)
-    in (updateReverseCount $ revs {
+        packageNodeIdMap = npm
+    }
+
+removePackage :: PackageId -> [PackageId]
+              -> ReverseIndex -> ReverseIndex
+removePackage pkgid deps ri@(ReverseIndex pindex revs nodemap) =
+    let index = PackageIndex.deletePackageId pkgid pindex
+        rd = delEdges (map (\d -> (nodemap ! d, [nodemap ! pkgid])) deps) revs
+    in ri {
         duplicatedIndex = index,
         reverseDependencies = rd,
-        flattenedReverse = packageNameClosure rd
-    }, rec)
-
-changePackage :: PackageId -> CombinedDeps -> CombinedDeps
-              -> ReverseIndex -> (ReverseIndex, Map PackageName [Version])
-changePackage pkgid deps deps' revs =
-    let index = PackageIndex.insert pkgid (duplicatedIndex revs)
-        allVersions = getAllVersions index
-        (rd,  rec)  = unregisterPackage allVersions pkgid deps (reverseDependencies revs)
-        (rd', rec') = registerPackage allVersions pkgid deps' rd
-    in (updateReverseCount $ revs {
+        packageNodeIdMap = nodemap
+    }
+
+changePackage :: PackageId -> [PackageId] -> [PackageId]
+              -> ReverseIndex -> ReverseIndex
+changePackage pkgid deps deps' ri@(ReverseIndex pindex revs nodemap) =
+    let index = PackageIndex.insert pkgid pindex
+        npm = Bimap.tryInsert pkgid (Bimap.size nodemap) nodemap
+        rd  = delEdges (map (\d -> (npm ! d, [npm ! pkgid])) deps) revs
+        rd' = insEdges (map (\d -> (npm ! d, [npm ! pkgid])) deps') rd
+    in ri {
         duplicatedIndex = index,
         reverseDependencies = rd',
-        flattenedReverse = packageNameClosure rd'
-    }, Map.unionWith union rec rec')
---------------------------------------------------------------------------------
--- Managing the RevDeps index.
-
-constructRevDeps :: PackageIndex PkgInfo -> RevDeps
-constructRevDeps index = foldl' (\revs pkg -> fst $ registerPackage (getAllVersions index) (packageId pkg) (getAllDependencies pkg) revs) Map.empty $ PackageIndex.allPackages index
+        packageNodeIdMap = npm
+    }
 
 getAllVersions :: Package pkg => PackageIndex pkg -> PackageName -> [Version]
 getAllVersions index = map packageVersion . PackageIndex.lookupPackageName index
```

Comment:
packageNodeIdMap is actually used a lot I think. It just isn't picked out by the accessor but rather pattern matched as `nodemap`.